Support for NATS as a subscriptions backplane

oising opened the pull request:
# NATS GraphQL Subscriptions backplane for HotChocolate

## What is NATS?

NATS is a simple, secure and high performance open source messaging system for cloud native applications, IoT messaging, and microservices architectures. It offers services for both asynchronous communication and streaming data.

## Why do I need NATS for my GraphQL Subscriptions?

If you scale your HotChocolate servers beyond a single node, some of your GraphQL Subscriptions will be handled by different servers. The subscriber may receive the subscription results on a different server than the publisher. In order to make this work, you need a backplane. Another example of a backplane is the Redis backplane that is included in HotChocolate. The Redis backplane is a good choice if you want to reuse some of your existing infrastructure. However, if you want to avoid the overhead of a Redis server, you can use the NATS backplane.

The NATS backplane is a simple and lightweight alternative to the Redis backplane. It is also typically faster than the Redis backplane, and cheaper, as you can run it on lightweight hardware, virtual or not, and scale it horizontally. The Publish/Subscribe pattern does not require a lot of memory, and Redis typically scales in a vertical fashion, adding more memory to a single server. NATS scales horizontally, adding more servers to the cluster.

## How do I use the NATS backplane?

You can start with a single node of NATS and see where you need to go from there. You can also start with a cluster of NATS servers and scale it. The NATS client uses a technique called client-side load balancing to distribute the load across the cluster. The NATS client will automatically discover the cluster and choose the correct server to publish to or subscribe from, based on a hash of the subject. The NATS client will also automatically reconnect to the cluster if a server goes down.

You do not need to enable persistence in the NATS server (JetStream) for Publish/Subscribe to function.

```csharp
using AlterNats;
using HotChocolate.Execution;
using HotChocolate.Subscriptions;

var builder = WebApplication.CreateBuilder(args);

builder.Services
    .AddNats(poolSize: 1, opts => opts with
    {
        Url = "nats://localhost:4222",
        // Optional serializer (defaults to System.Text.Json)
        Serializer = new MessagePackNatsSerializer()
    })
    .AddNatsSubscriptions()
    .AddGraphQLServer()
    .AddMutationConventions()
    .AddQueryType<Query>()
    .AddMutationType<Mutation>()
    .AddSubscriptionType<Subscription>();

var app = builder.Build();

app.UseWebSockets();
app.UseRouting();
app.UseEndpoints(ep =>
{
    ep.MapGraphQL();
});

await app.RunAsync();
```

> **NOTE**:
> _You need to target net6.0 or higher to use this provider, as the NATS client dependency is not available for net5.0 or lower._ 

## Multiple Environment Support

If you wish to share a NATS instance/cluster with multiple, distinct GraphQL servers, you must provide a unique subject prefix for this instance, e.g. an environment string like "dev01". This prefix is passed in the initial setup, like so: `AddNatsSubscription(prefix: "dev01")`. All servers with the same prefix will be part of the same publish/subscribe group, and will receive all messages. The prefix will be prepended to all NATS subjects with the "." subject token separator, and as such can be monitored with the NATS CLI using the &gt; operator: 

```bash
./nats -s localhost:4222 subscribe "dev01.>"
```

If you do not provide a prefix, the server will use the default prefix of "graphql". Only a-z, A-Z and 0-9 characters are permitted. The prefix is case sensitive.

## Technical Information about NATS

NATS has a nice command line interface which you can use to monitor the cluster. You can also use the NATS dashboard to monitor the cluster.

### NATS PubSub overview using NATS CLI

https://www.youtube.com/watch?v=jLTVhP08Tq0

michaelstaib commented:
Is there a docker image with this available?

michaelstaib commented:
https://hub.docker.com/_/nats

michaelstaib commented:
We need to create a squadron package for this so we can test this thing.

PascalSenn commented on src/HotChocolate/Core/src/Subscriptions.Nats/HotChocolate.Subscriptions.Nats.csproj:
```diff
@@ -0,0 +1,21 @@
+    </ItemGroup>
+
+    <ItemGroup>
+      <PackageReference Include="AlterNats.Hosting" Version="1.0.5" />
```
Is the reason to use this client and not the offical one performance?

oising replied:
Yeah, the official package is really awful -- it's a straight port of the Go version, and doesn't use any of the normal dotnet idioms. AlterNats is zero-alloc and very modern, has more functionality, and is following all the idioms with async that you'd want to see.

oising commented:
> We need to create a squadron package for this so we can test this thing.

Yeah, I looked to see what Squadron was, and it seems there is no OOTB NATS package - if this is something that is expected to be written by us* (i.e. could be me?) then let me know where I could start.

oising commented:
Waiting on https://github.com/SwissLife-OSS/squadron/pull/123

oising commented:
Uncovered a race condition in my NatsPubSub implementation -- marking as draft until squadron tests can prove otherwise

oising commented:
@PascalSenn @michaelstaib - ready to review now, but there's still an issue in core with net7-rc1

michaelstaib commented:
Let me check

oising commented:
Added multiple environment support:
- updated readme.md
- updated top comment above

## Multiple Environment Support

If you wish to share a NATS instance/cluster with multiple, distinct GraphQL servers, you must provide a unique subject prefix for this instance, e.g. and environment string like "dev01". All servers with the same prefix will be part of the same publish/subscribe group, and will receive all messages. The prefix will be prepended to all NATS subjects with the "." subject token separator, and as such can be monitored with the NATS CLI using the &gt; operator: 

```bash
./nats -s localhost:4222 subscribe "dev01.>"
```

If you do not provide a prefix, the server will use the default prefix of "graphql". Only a-z, A-Z and 0-9 characters are permitted. The prefix is case sensitive.

michaelstaib commented:
Before we can move forward on this we have to solve the connection issue that we have on the local dev system. I have started investigating.

oising commented:
Pushed new updates to deal with topics that are complex objects; previously I was assuming topics would only be strings. Topics are now converted into a 128-bit hash of a MessagePack serialized byte stream. This also happens for plain strings, as they may contain characters that are not valid for a NATS subject string. Main changes:

- All messages are now wrapped in a NatsMessageEnvelope, which has a field to convey the type of message: a fully body/event, or a terminator message to signal to end of the stream. This is more reliable than a magic string. 
- The hashing algorithm being used is SpookyHash, an updated version of the Jenkins Hash algorithm; which is a non-cryptographic has designed for speed and extreme collision avoidance. Crypto-based hashing is slow, and has a different threat model.

michaelstaib commented:
We will be waiting until https://github.com/Cysharp/AlterNats/issues/23 is resolved.

oising commented:
@PascalSenn @michaelstaib This should be unblocked now, but you'll need to publish a new squadron package with my fix and then we can update this PR to use it

https://github.com/SwissLife-OSS/squadron/pull/126

oising commented:
> @PascalSenn @michaelstaib This should be unblocked now, but you'll need to publish a new squadron package with my fix and then we can update this PR to use it
> 
> [SwissLife-OSS/squadron#126](https://github.com/SwissLife-OSS/squadron/pull/126)

Squadron published this morning 18-preview.4 with the osx fix. I've just merged into this PR. We're officially ready to go.
